src/ray/raylet/worker_pool.cc

```diff
@@ -739,27 +766,31 @@ void WorkerPool::PushWorker(const std::shared_ptr<WorkerInterface> &worker) {
   RAY_CHECK(worker->GetAssignedTaskId().IsNil())
       << "Idle workers cannot have an assigned task ID";
   auto &state = GetStateForLanguage(worker->GetLanguage());
-  bool found;
-  bool used;
-  TaskID task_id;
-  InvokePopWorkerCallbackForProcess(state.starting_dedicated_workers_to_tasks,
-                                    worker->GetShimProcess(), worker, PopWorkerStatus::OK,
-                                    &found, &used, &task_id);
-  if (found) {
-    // The worker is used for the actor creation task with dynamic options.
-    if (!used) {
-      // Put it into idle dedicated worker pool.
-      // TODO(guyang.sgy): This worker will not be used forever. We should kill it.
-      state.idle_dedicated_workers[task_id] = worker;
+
+  // Try to dispatch a task to this worker.
+  bool worker_used = false;
+  auto runtime_env_hash = worker->GetRuntimeEnvHash();
+  RAY_LOG(DEBUG) << "PushWorker with runtime env " << runtime_env_hash;
+  auto &waiting_tasks = state.waiting_tasks_by_env_hash[runtime_env_hash];
+  while (!worker_used && !waiting_tasks.empty()) {
+    auto &task_info = waiting_tasks.front();
+    worker_used = task_info.callback(worker, PopWorkerStatus::OK);
```

Comment:
Hmm why would this ever fail? Seems weird we have this while loop here.

Reply:
This is the same as before. We do this because tasks can get canceled by the cluster_task_manager, so we check to make sure that the worker is actually being used.